File: services/academic-assessment-service/src/main/java/com/visor/school/assessment/event/ReportEventPublisher.java
package com.visor.school.assessment.event;

import com.visor.school.common.events.BaseEvent;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.stereotype.Component;

import java.time.Instant;
import java.util.UUID;

/**
 * Event publisher for report collection events
 */
@Component
public class ReportEventPublisher {
    private static final String EXCHANGE_NAME = "academic_events";
    
    private final RabbitTemplate rabbitTemplate;

    public ReportEventPublisher(RabbitTemplate rabbitTemplate) {
        this.rabbitTemplate = rabbitTemplate;
    }

    /**
     * Publish ReportCollectedEvent when class teacher collects exam results
     */
    public void publishReportCollectedEvent(
            UUID collectionId,
            UUID classId,
            UUID classTeacherId,
            String academicYear,
            String term) {
        ReportCollectedEvent event = new ReportCollectedEvent(
            collectionId,
            classId,
            classTeacherId,
            academicYear,
            term
        );

        rabbitTemplate.convertAndSend(EXCHANGE_NAME, "report.collected", event);
    }

    /**
     * Publish ReportSubmittedEvent when class teacher submits report to school
     */
    public void publishReportSubmittedEvent(
            UUID submissionId,
            UUID collectionId,
            UUID classId,
            UUID classTeacherId) {
        ReportSubmittedEvent event = new ReportSubmittedEvent(
            submissionId,
            collectionId,
            classId,
            classTeacherId
        );

        rabbitTemplate.convertAndSend(EXCHANGE_NAME, "report.submitted", event);
    }
}

/**
 * Report Collected Event
 */
class ReportCollectedEvent extends BaseEvent {
    private final UUID collectionId;
    private final UUID classId;
    private final UUID classTeacherId;
    private final String academicYear;
    private final String term;

    public ReportCollectedEvent(UUID collectionId, UUID classId, UUID classTeacherId,
                               String academicYear, String term) {
        super(UUID.randomUUID(), Instant.now(), "1.0");
        this.collectionId = collectionId;
        this.classId = classId;
        this.classTeacherId = classTeacherId;
        this.academicYear = academicYear;
        this.term = term;
    }

    @Override
    public String getEventType() {
        return "report.collected";
    }

    @Override
    public UUID getAggregateId() {
        return collectionId;
    }

    @Override
    public String getAggregateType() {
        return "ExamResultCollection";
    }

    // Getters
    public UUID getCollectionId() { return collectionId; }
    public UUID getClassId() { return classId; }
    public UUID getClassTeacherId() { return classTeacherId; }
    public String getAcademicYear() { return academicYear; }
    public String getTerm() { return term; }
}

/**
 * Report Submitted Event
 */
class ReportSubmittedEvent extends BaseEvent {
    private final UUID submissionId;
    private final UUID collectionId;
    private final UUID classId;
    private final UUID classTeacherId;

    public ReportSubmittedEvent(UUID submissionId, UUID collectionId, UUID classId, UUID classTeacherId) {
        super(UUID.randomUUID(), Instant.now(), "1.0");
        this.submissionId = submissionId;
        this.collectionId = collectionId;
        this.classId = classId;
        this.classTeacherId = classTeacherId;
    }

    @Override
    public String getEventType() {
        return "report.submitted";
    }

    @Override
    public UUID getAggregateId() {
        return submissionId;
    }

    @Override
    public String getAggregateType() {
        return "ReportSubmission";
    }

    // Getters
    public UUID getSubmissionId() { return submissionId; }
    public UUID getCollectionId() { return collectionId; }
    public UUID getClassId() { return classId; }
    public UUID getClassTeacherId() { return classTeacherId; }
}

File: services/academic-assessment-service/src/main/kotlin/com/visor/school/assessment/event/ReportEventPublisher.kt
package com.visor.school.assessment.event

import com.visor.school.common.events.BaseEvent
import org.springframework.amqp.rabbit.core.RabbitTemplate
import org.springframework.stereotype.Component
import java.time.Instant
import java.util.UUID

/**
 * Event publisher for report collection events
 */
@Component
class ReportEventPublisher(
    private val rabbitTemplate: RabbitTemplate
) {
    private val exchangeName = "academic_events"

    /**
     * Publish ReportCollectedEvent when class teacher collects exam results
     */
    fun publishReportCollectedEvent(
        collectionId: UUID,
        classId: UUID,
        classTeacherId: UUID,
        academicYear: String,
        term: String
    ) {
        val event = ReportCollectedEvent(
            collectionId = collectionId,
            classId = classId,
            classTeacherId = classTeacherId,
            academicYear = academicYear,
            term = term
        )

        rabbitTemplate.convertAndSend(exchangeName, "report.collected", event)
    }

    /**
     * Publish ReportSubmittedEvent when class teacher submits report to school
     */
    fun publishReportSubmittedEvent(
        submissionId: UUID,
        collectionId: UUID,
        classId: UUID,
        classTeacherId: UUID
    ) {
        val event = ReportSubmittedEvent(
            submissionId = submissionId,
            collectionId = collectionId,
            classId = classId,
            classTeacherId = classTeacherId
        )

        rabbitTemplate.convertAndSend(exchangeName, "report.submitted", event)
    }
}

/**
 * Report Collected Event
 */
data class ReportCollectedEvent(
    val collectionId: UUID,
    val classId: UUID,
    val classTeacherId: UUID,
    val academicYear: String,
    val term: String,
    val eventIdOverride: UUID = UUID.randomUUID(),
    val eventTimestamp: Instant = Instant.now(),
    val eventVersion: String = "1.0"
) : BaseEvent(eventIdOverride, eventTimestamp, eventVersion) {
    override val eventType: String = "report.collected"
    override fun getAggregateId(): UUID = collectionId
    override fun getAggregateType(): String = "ExamResultCollection"
}

/**
 * Report Submitted Event
 */
data class ReportSubmittedEvent(
    val submissionId: UUID,
    val collectionId: UUID,
    val classId: UUID,
    val classTeacherId: UUID,
    val eventIdOverride: UUID = UUID.randomUUID(),
    val eventTimestamp: Instant = Instant.now(),
    val eventVersion: String = "1.0"
) : BaseEvent(eventIdOverride, eventTimestamp, eventVersion) {
    override val eventType: String = "report.submitted"
    override fun getAggregateId(): UUID = submissionId
    override fun getAggregateType(): String = "ReportSubmission"
}
